tests/lambdas/leverDev.test.js

```diff
@@ -0,0 +1,79 @@
+const lambda = require('../../src/functions/lever')
+const got = require('got')
+const url = require('url')
+
+jest.mock('got')
+jest.mock('url')
+
+describe('different data, not production', () => {
+  let response
+
+  beforeAll(async () => {
+    url.URL.mockReturnValueOnce({
+      href: 'http://localhost:8000/meta.json',
+      origin: 'http://localhost:8000',
+      protocol: 'http:',
+      username: '',
+      password: '',
+      host: 'localhost:8000',
+      hostname: 'localhost',
+      port: '8000',
+      pathname: '/meta.json',
+      search: '',
+      hash: ''
+    })
+
+    url.URL.mockReturnValueOnce({
+      href: 'https://api.lever.co/v0/postings/yld?mode=json',
+      origin: 'https://api.lever.co',
+      protocol: 'https:',
+      username: '',
+      password: '',
+      host: 'api.lever.co',
+      hostname: 'api.lever.co',
+      port: '',
+      pathname: '/v0/postings/yld',
+      search: '?mode=json',
+      hash: ''
+    })
```

Comment:
Here we don't need to mock the URL module, its a native JS module so we can rely on it to give us what we want. 
Using this mock here means that we aren't making sure we're passing the URL module the correct values, because we're in control of what we pass it, it's important to make sure we test it. 

In my version you can see we're testing that the got module is called with the correct object from the URL as the first argument. The reason we do it like this is that if there are changes in the lever function to the got requests and urls then the tests will fail. 

[We set up the URLS](https://github.com/yldio/yld.io/compare/feat/lever-unit-tests?expand=1#diff-4772b57a44b209349e50cff83b53d1c8R17-R18)
[then we make sure that the got mock is called with the right arguments](https://github.com/yldio/yld.io/compare/feat/lever-unit-tests?expand=1#diff-4772b57a44b209349e50cff83b53d1c8R59-R62)

Reply:
So at which point is it "safe" to just use the modules as-is? You said "native JS module, so we can rely on it". Would you say the same about, for example, lodash? Or "any module that just transforms data (as opposed to fetching / sending) is safe to use as-is"? 

> Using this mock here means that we aren't making sure we're passing the URL module the correct values, because we're in control of what we pass it, it's important to make sure we test it.

But isn't mocking about faking the data we want to receive? I'm also having trouble parsing the above sentence - do all the instances of "it" apply to the "URL module".

Reply:
No lodash is just a library of scripts whereas URL is a native module to JS. We can trust the native module more than something like lodash, although in this unit test we're not mocking the sort function of lodash.

Mocking is about isolating the object under test and making sure that the internals of that function perform the tasks we want them to. The key point here is `isolation`. If you mock the dependencies you know the test is only testing the internals, not externals. Another way of testing this lambda is to use `nock` to return the correct data from an HTTP request, but that means we're still relying on Got to do the request and formulate the http request correctly, which is not what we want to test. 

Mocking in unit testing allows us to control all of the external dependencies of a function. E.g. the got module or even a complex data transformation function that we have written ourselves. Remember, we don't want to test that functionality, we only want to test our module.

If you've not heard of the testing pyramid I suggest reading up on it, Martin Fowler originated the idea I believe: 
![image](https://user-images.githubusercontent.com/7108120/63075683-0e0f5a00-bf65-11e9-9f6f-84e037d7fcfb.png)